packages/react-native-gesture-handler/src/v3/createNativeWrapper.tsx

```diff
@@ -0,0 +1,83 @@
+import React, { useImperativeHandle, useRef } from 'react';
+
+import { NativeWrapperProps } from './hooks/utils';
+import { useNativeGesture } from './hooks/gestures';
+import { NativeDetector } from './detectors/NativeDetector';
+import type { NativeWrapperProperties } from './types/NativeWrapperType';
+import { NativeGesture } from './hooks/gestures/native/useNativeGesture';
+import { DetectorType, InterceptingGestureDetector } from './detectors';
+import { VirtualDetector } from './detectors/VirtualDetector/VirtualDetector';
+
+export type ComponentWrapperRef<P, T> = {
+  componentRef?: React.ComponentType<P> & T;
+  gestureRef?: NativeGesture;
+};
+
+export default function createNativeWrapper<P>(
+  Component: React.ComponentType<P>,
+  config: Readonly<NativeWrapperProperties> = {},
+  detectorType: DetectorType = DetectorType.Native
+) {
+  const ComponentWrapper = (
+    props: P & NativeWrapperProperties & { ref?: React.RefObject<unknown> }
+  ) => {
+    // Filter out props that should be passed to gesture handler wrapper
+    const { gestureHandlerProps, childProps } = Object.keys(props).reduce(
+      (res, key) => {
+        // @ts-ignore TS being overly protective with it's types, see https://github.com/microsoft/TypeScript/issues/26255#issuecomment-458013731 for more info
+        if (NativeWrapperProps.has(key)) {
+          // @ts-ignore FIXME(TS)
+          res.gestureHandlerProps[key] = props[key];
+        } else {
+          // @ts-ignore FIXME(TS)
+          res.childProps[key] = props[key];
+        }
+        return res;
+      },
+      {
+        gestureHandlerProps: { ...config }, // Watch out not to modify config
+        childProps: {
+          enabled: props.enabled,
+          hitSlop: props.hitSlop,
+        } as P,
+      }
+    );
+
+    if (gestureHandlerProps.disableReanimated === undefined) {
+      gestureHandlerProps.disableReanimated = true;
+    }
+
+    const native = useNativeGesture(gestureHandlerProps);
+
+    const componentRef = useRef<React.ComponentType<P>>(null);
+    const gestureRef = useRef<NativeGesture>(native);
+
+    useImperativeHandle(props.ref, () => ({
+      componentRef: componentRef.current,
+      gestureRef: gestureRef.current,
+    }));
```

Comment:
I'm not sure if this is at all useful:
1. It's a breaking change compared to the previous api and the components stop being a drop-in replacement - `scrollRef.current.scrollTo` becomes `scrollRef.current.componentRef.scrollTo`.
2. Is `gestureRef` actually usable? There's no way to react to changes happening to it in an easy way. It will be `null` after creating the ref, and that's what will be used for relations during the first render. To actually read the ref, something would need to cause a render after it's set. If it happens to change (like in the case of child remount), there's no signal that the parent (gesture) should react to it.

Reply:
Hmm, would it be possible to split this? Maybe it would be possible to keep `ref` as a view ref and add a callback like `onGestureUpdate` which would run every time the gesture changes? This way, it's reactive, so you don't have to do any tricks to trigger updates.

Internally, we only care about the handler tag, which doesn't change so I don't think we care about any updates, just the first time it's set.

Reply:
Let me know what you think about [a66158c](https://github.com/software-mansion/react-native-gesture-handler/pull/3800/commits/a66158c6ef96304730f6ce1387220cc169991d74)